File: app/src/main/java/com/example/myapplication/presentation/ui/adapter/UserAdapter.java
package com.example.myapplication.presentation.ui.adapter;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;

import com.example.myapplication.R;
import com.example.myapplication.databinding.ItemUserBinding;
import com.example.myapplication.domain.model.User;

public class UserAdapter extends ListAdapter<User, UserAdapter.UserViewHolder> {

    public interface OnUserClickListener {
        void onUserClick(User user);
    }

    public interface OnUserLongClickListener {
        void onUserLongClick(User user);
    }

    private static final DiffUtil.ItemCallback<User> DIFF_CALLBACK = new DiffUtil.ItemCallback<User>() {
        @Override
        public boolean areItemsTheSame(@NonNull User oldItem, @NonNull User newItem) {
            return oldItem.getId() == newItem.getId();
        }

        @Override
        public boolean areContentsTheSame(@NonNull User oldItem, @NonNull User newItem) {
            return oldItem.equals(newItem);
        }
    };

    private final OnUserClickListener onItemClick;
    private final OnUserLongClickListener onItemLongClick;

    public UserAdapter(OnUserClickListener onItemClick, OnUserLongClickListener onItemLongClick) {
        super(DIFF_CALLBACK);
        this.onItemClick = onItemClick;
        this.onItemLongClick = onItemLongClick;
    }

    @NonNull
    @Override
    public UserViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        return new UserViewHolder(ItemUserBinding.inflate(inflater, parent, false));
    }

    @Override
    public void onBindViewHolder(@NonNull UserViewHolder holder, int position) {
        holder.bind(getItem(position));
    }

    class UserViewHolder extends RecyclerView.ViewHolder {

        private final ItemUserBinding binding;

        UserViewHolder(ItemUserBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

        void bind(User user) {
            binding.textUserName.setText(user.getName());
            binding.textUserId.setText(binding.getRoot().getContext().getString(R.string.user_id_format, user.getId()));
            binding.textUserAge.setText(binding.getRoot().getContext().getString(R.string.user_age_format, user.getAge()));
            binding.textUserWeight.setText(binding.getRoot().getContext().getString(R.string.user_weight_format, user.getWeight()));

            binding.getRoot().setOnClickListener(v -> onItemClick.onUserClick(user));
            binding.getRoot().setOnLongClickListener(v -> {
                onItemLongClick.onUserLongClick(user);
                return true;
            });
        }
    }
}

File: app/src/main/java/com/example/myapplication/presentation/ui/adapter/UserAdapter.kt
package com.example.myapplication.presentation.ui.adapter

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.example.myapplication.R
import com.example.myapplication.databinding.ItemUserBinding
import com.example.myapplication.domain.model.User

class UserAdapter(
    private val onItemClick: (User) -> Unit,
    private val onItemLongClick: (User) -> Unit,
) : ListAdapter<User, UserAdapter.UserViewHolder>(DiffCallback) {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): UserViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        return UserViewHolder(ItemUserBinding.inflate(inflater, parent, false))
    }

    override fun onBindViewHolder(holder: UserViewHolder, position: Int) {
        holder.bind(getItem(position))
    }

    inner class UserViewHolder(
        private val binding: ItemUserBinding,
    ) : RecyclerView.ViewHolder(binding.root) {

        fun bind(user: User) {
            binding.textUserName.text = user.name
            binding.textUserId.text = binding.root.context.getString(R.string.user_id_format, user.id)
            binding.textUserAge.text = binding.root.context.getString(R.string.user_age_format, user.age)
            binding.textUserWeight.text =
                binding.root.context.getString(R.string.user_weight_format, user.weight)

            binding.root.setOnClickListener { onItemClick(user) }
            binding.root.setOnLongClickListener {
                onItemLongClick(user)
                true
            }
        }
    }

    private object DiffCallback : DiffUtil.ItemCallback<User>() {
        override fun areItemsTheSame(oldItem: User, newItem: User): Boolean = oldItem.id == newItem.id

        override fun areContentsTheSame(oldItem: User, newItem: User): Boolean = oldItem == newItem
    }
}
